File: src/main/java/no/lambdaconsulting/javatokotlin/dao/PetDaoJdbcTemplate.java
package no.lambdaconsulting.javatokotlin.dao;

import no.lambdaconsulting.javatokotlin.domain.PetSummary;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public class PetDaoJdbcTemplate implements PetDao {

    private final JdbcTemplate jdbcTemplate;
    private final RowMapper<PetSummary> rowMapper = (res, rowNum) ->
        new PetSummary(
            res.getInt("id"),
            res.getString("name"),
            res.getInt("breed_id")
        );

    public PetDaoJdbcTemplate(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    @Override
    public List<PetSummary> findByPerson(int personId) {
        return jdbcTemplate.query(
            "SELECT id, name, breed_id FROM pet where person_id = ?",
            new Object[]{personId},
            rowMapper
        );
    }

    @Override
    public PetSummary getById(int petId) {
        return jdbcTemplate.queryForObject(
            "SELECT id, name, breed_id FROM pet WHERE id = ?",
            new Object[]{petId},
            rowMapper
        );
    }
}

File: src/main/kotlin/no/lambdaconsulting/javatokotlin/dao/PetDaoJdbcTemplate.kt
package no.lambdaconsulting.javatokotlin.dao

import no.lambdaconsulting.javatokotlin.domain.PetSummary
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Repository

interface PetDao {
    fun getById(petId: Int): PetSummary?
    fun findByPerson(personId: Int): List<PetSummary>
}

@Repository
class PetDaoJdbcTemplate(private val jdbcTemplate: JdbcTemplate) : PetDao {

    override fun getById(petId: Int): PetSummary? =
        /* FIX> 2nd param had to be be added, otherwise the SQL statement had no value for the placeholder */
        jdbcTemplate.queryForObject("SELECT id, name, breed_id FROM pet WHERE id = ?",
            arrayOf(petId)) { rs, _ ->
            PetSummary(rs.getInt("id"), rs.getString("name"), rs.getInt("breed_id"))
        }

    override fun findByPerson(personId: Int): List<PetSummary> =
        /* FIX> 2nd param had to be be added, otherwise the SQL statement had no value for the placeholder */
        jdbcTemplate.query("SELECT id, name, breed_id FROM pet where person_id = ?",
            arrayOf(personId)) { rs, _ ->
            PetSummary(rs.getInt("id"), rs.getString("name"), rs.getInt("breed_id"))
        }
}